CenterForOpenScience/osf.io
# [ENG-649] Figshare Article Types Update - Phase One

## Purpose

figshare used to only let type 4 `fileset` behave like folders, i.e. it can contain multiple files.  However, it suddenly changed recently without any announcement that 1) all article types are now folder-like, that 2) type 4 `fileset` is unofficially deprecated and converted to type 3 `dataset` internally, and that 3) new types are added.

This phase one PR simply refactors `fileset` to `dataset` without touching other existing types (except for name changes). The phase two, currently documented as a future [SVCS task](https://openscience.atlassian.net/browse/ENG-996), aims to bring the figshare add-on fully up-to-date.

- [x] ~~Fix tests if any~~ Travis issue, fixed.
- [x] Create a ticket.
- [x] Update JIRA Page link when done.

## Changes

### Primary Change

* Now uses a bi-directional dictionary for `defined_type` and `defined_type_name` mapping.
* Added a set for folder-like types (currently both `dataset` and `fileset`). Everywhere we used to check for `fileset` or `4`, we now check set membership.

### By-products

* Added new types and renamed existing ones.
* Did some minor functionality-preserving code changes and style updates to 1) avoid lines that are too long, to 2) improve `if` statement by replacing equality check with set membership check and to 3) improve style consistency.

## QA Notes

Please verify that both projects and folders (article type 3 and 4) are listed and can be set as OSF project storage root during figshare add-on configuration.

figshare changed their data model and tweaked the API slightly so you may not be able to see a working figshare without the WB side fix ([WB-PR#378](https://github.com/CenterForOpenScience/waterbutler/pull/378) and [ENG-Ticket#380](https://openscience.atlassian.net/browse/ENG-380)). Although these two PRs don't block each other, I would suggest them be merged, tested and released together.

In addition, if you are confused on recent weird figshare behaviors (as I was for quite a while), here is a [JIRA page](https://openscience.atlassian.net/wiki/spaces/SVCS/pages/817037313) that may help.

## Documentation

Not for the OSF and API.

## Side Effects

No

## Ticket

https://openscience.atlassian.net/browse/ENG-649
